minimize_neldermead() stop at user requested maxiter or maxfev

samtygier opened the pull request:
If the user sets maxiter, then don't stop when the default maxfev is
reached, and vice versa. Only use the defaults if neither maxiter or
maxfev has been set.

Fixes #5966

Changes from https://github.com/scipy/scipy/pull/5968
Use np.inf to signify no limit on maxiter or maxfev
Add documentation

The following test program shows the change for a selection of function calls with a slow to converge function.
```
#!/usr/bin/env python
import numpy as np
from numpy import sqrt, sin, arctan2
import scipy
import scipy.optimize
print scipy.__version__

def test_func(v):
	x,y = v
	r,t = sqrt(x**2+y**2), arctan2(x,y)
	a,b  = 20, 0.5
	pen = sin(r*a + t)+r*b
	return pen

opts = [{},
{"maxiter":5000},
{"maxfev":50},
{"maxfev":5000},
{"maxiter":2000, "maxfev":2000},
{"maxiter":5000, "maxfev":5000},
{"maxiter":5000, "maxfev":None},
{"maxiter":5000, "maxfev":np.inf},
]

print "maxiter\tmaxfev\titers\tfevs\tsuccess"
for n, opt in enumerate(opts):
	res = scipy.optimize.minimize(test_func, [1,1], method='Nelder-Mead', options=opt)
	print opt.get("maxiter", "-"),"\t",opt.get("maxfev","-"),"\t", res["nit"],"\t", res["nfev"],"\t", res["success"]
```

With 0.17.0
```
maxiter maxfev  iters   fevs    success
-       -       217     400     False
5000    -       217     400     False
-       50      26      50      False
-       5000    400     731     False
2000    2000    1092    2000    False
5000    5000    2114    3902    True
5000    None    217     400     False
5000    inf     2114    3902    True
```
With the patch i get 
```
maxiter maxfev  iters   fevs    success
-       -       217     400     False
5000    -       2114    3902    True
-       50      26      50      False
-       5000    2114    3902    True
2000    2000    1092    2000    False
5000    5000    2114    3902    True
5000    None    2114    3902    True
5000    inf     2114    3902    True
```

The only changes are that if the user has requested a large maxiter, or a large maxfev, the minimizer will now run to that number, rather than stopping early. This makes it more likely to reach convergence, but may cause it to take more time.

pv commented:
Needs tests --- might also be of interest to check whether the other solvers that have maxfev are consistent with this.

samtygier commented:
Here is a test for a few combinations of arguments and starting vectors. I could make it more generic and test all the algorithms. Is it ok to just trust the nfev returned by the optimization function, or should I put a counter in the func?

Powell is the only other method which takes both maxfev and maxiter. So I think a similar fix would be good for it as well, once you are happy with the approach.

pv commented:
Test looks ok, although might be more robust to count the nfev in the function itself.

The PR needs rebase.

samtygier commented:
Rebased. Also counting the function iterations in the function itself and assert that it is equal to the returned value.

pv commented on scipy/optimize/optimize.py:
```diff
@@ -490,10 +497,15 @@ def _minimize_neldermead(func, x0, args=(), callback=None,
+    elif maxiter is not None and maxiter != np.inf:
+        maxfun = np.inf
+    elif maxfun is not None and maxfun != np.inf:
+        maxiter = np.inf
```
The logic seems a bit wrong here:  for `maxiter=None, maxfev=np.inf`, this ends up running with `maxiter=None` which probably ends up with None vs. int comparison, which raises an exception.

Probably best to just write
```
if maxiter is None and maxfun is None:
    ...
elif maxiter is None:
    maxiter = np.inf
elif maxfun is None:
    maxfun = np.inf
```

samtygier replied:
`maxiter=None, maxfun=np.inf` is a tricky one. That will turn it into `maxiter=np.inf, maxfun=np.inf`, which might result in an infinite loop if there is some function that does not converge.

My (somewhat awkward) version was trying to avoid the case setting them both to inf, i.e. only set maxfun to inf if maxiter has a finite value.

Maybe it should interpret `maxiter=None, maxfun=np.inf`, as "never stop on maxfun, but use the default for maxiter", then it would become `maxiter=N*200, maxfun=np.inf`.

```

    # If neither are set, then set both to default
    if maxiter is None and maxfun is None:
        maxiter = N * 200
        maxfun = N * 200
    # Convert remaining Nones, to np.inf, unless the other is np.inf, in
    # which case use the default to avoid unbounded iteration
    elif maxiter is None:
        if maxfun is np.inf:
            maxiter = N * 200
        else:
            maxiter = np.inf
    elif maxfun is None:
        if maxiter is np.inf:
            maxfun = N * 200
        else:
            maxfun = np.inf

```

I am not 100% sure which is best.

If the users requests just maxfun=5000, then I think the code should respect that and not stop earlier (as it originally did). In terms of negative effects on existing code, this could make them take longer, but it would still definitely terminate.

If the user requests just maxfun=np.inf, should that be taken literally, or should the user be protected against infinite loops? The old code protected them. If they really want iteration to never stop without convergence they can set both maxfun and maxiter to np.inf.

(Either way, for the fixed version, should I 1) just add a new commit to this branch, 2) make a new branch and new pull request, 3) overwrite this branch with a push --force?).

ev-br replied:
1) or 3) please.

samtygier commented:
Pushed a new version to my branch as requested. This now handles the None,np.inf and np.inf,None cases by treating them as `N*200,np.inf` and `np.inf,N*200 `respectively. 

I also updated the test to include the inf cases, and to fix a couple of typos.

So updating the table of outcomes, with current git master:
```
0.18.0.dev0+7e807cc
maxiter maxfev  iters   fevs    success
-       -       217     400     False
5000    -       217     400     False
-       50      26      50      False
-       5000    400     731     False
2000    2000    1092    2000    False
5000    5000    2114    3902    True
5000    None    217     400     False
5000    inf     2114    3902    True
inf     None    217     400     False
None    inf     400     731     False
inf     inf     2114    3902    True
```

and with this patch

```
maxiter maxfev  iters   fevs    success
-       -       217     400     False
5000    -       2114    3902    True
-       50      26      50      False
-       5000    2114    3902    True
2000    2000    1092    2000    False
5000    5000    2114    3902    True
5000    None    2114    3902    True
5000    inf     2114    3902    True
inf     None    217     400     False
None    inf     400     731     False
inf     inf     2114    3902    True
```

pv commented on scipy/optimize/optimize.py:
```diff
@@ -490,10 +497,22 @@ def _minimize_neldermead(func, x0, args=(), callback=None,
+    # Convert remaining Nones, to np.inf, unless the other is np.inf, in
+    # which case use the default to avoid unbounded iteration
+    elif maxiter is None:
+        if maxfun is np.inf:
```
`maxfun == np.inf`, floats are not singletons `float('inf') is np.inf == False`

pv commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -1134,6 +1134,54 @@ def test_brute(self):
+                self.funcalls = 0
+                res = optimize.minimize(self.slow_func, start_v,
+                      method='Nelder-Mead', options={"maxfev":mfev})
+                assert(self.funcalls == res["nfev"])
```
Use `assert_` from `numpy.testing` (plain assert is removed by Python when running in "optimized" mode)

pv commented on scipy/optimize/optimize.py:
```diff
@@ -490,10 +497,22 @@ def _minimize_neldermead(func, x0, args=(), callback=None,
         maxiter = N * 200
-    if maxfun is None:
         maxfun = N * 200
+    # Convert remaining Nones, to np.inf, unless the other is np.inf, in
```
Empty line before comment

samtygier commented:
Thanks for comments, fixed in new version.

samtygier commented:
Are there any more comments? Can the needs-work label be removed?

nmayorov commented:
Two suggestions:

1. Handle http://scipy.github.io/devdocs/optimize.minimize-powell.html in the similar fashion.
2. Maybe instead of adding a Notes section it's better to combine the two parameters as `maxiter, maxnfev` and explain how they both going to work in one place. But this is just my preference.

I checked other solvers in `minimize`: `L-BFGS-B` have `maxiter` and `maxnfev`, but they are set to fixed values (perhaps not optimal decision), so not necessary to change. Other solver don't have both `maxiter` and `maxnfev`.

samtygier commented:
I have added the same patch fix for powell (i think its the only other function that needs this), and updated the test to check powell as well. Also fixed the condition for last part of the test.

I got rid of the notes section and move the new details into the argument section as that is probably is the clearest way. I have not merged the 2 parameters, as its simpler to still have them separate, but maybe its worth merging them to reduce the repetition. Happy to do this if requested.

nmayorov commented:
maxiter : int
 -        Maximum number of iterations to perform.
 +        Maximum number of iterations to perform. Will default to N*200,
 +        where ``N`` is the dimension, if neither maxiter or maxfev is
 +        set. If both maxiter and maxfev are set, minimization will stop
 +        at the first reached.
      maxfev : int
 -        Maximum number of function evaluations to make.
 +        Maximum number of function evaluations to make. Will default to
 +        N*200, where ``N`` is the dimension, if neither maxiter or
 +        maxfev is set.

samtygier commented:
Ok, I have used the merged parameters. Should I also add to the release notes? Something like:

Nelder-Mead and Powell minimization will now only set defaults for
maximum iterations or function evaluations if neither limit is set by
the caller. In some cases with a slow converging function and only 1
limit set, the minimization may continue for longer than with previous
versions and so is more likely to reach convergence. See issue gh-5966.

ev-br commented:
Yeah, please add it.

ev-br commented:
Ok, let's get this in. Thanks @samtygier, all
